route.go

```diff
@@ -147,53 +188,108 @@ func (route *baseRoute) Shutdown() error {
 }
 
 // to view the state of the table/route at any point in time
-func (route *RouteSendAllMatch) Snapshot() RouteSnapshot {
+func makeSnapshot(route *baseRoute, routeType string) RouteSnapshot {
 	conf := route.config.Load().(RouteConfig)
-	dests := make([]*Destination, len(conf.Dests))
-	for i, d := range conf.Dests {
+	dests := make([]*Destination, len(conf.Dests()))
+	for i, d := range conf.Dests() {
 		dests[i] = d.Snapshot()
 	}
-	return RouteSnapshot{conf.Matcher, dests, "sendAllMatch", route.key}
+	return RouteSnapshot{*conf.Matcher(), dests, routeType, route.key}
+
+}
+
+func (route *RouteSendAllMatch) Snapshot() RouteSnapshot {
+	return makeSnapshot(&route.baseRoute, "sendAllMatch")
 }
 
 func (route *RouteSendFirstMatch) Snapshot() RouteSnapshot {
-	conf := route.config.Load().(RouteConfig)
-	dests := make([]*Destination, len(conf.Dests))
-	for i, d := range conf.Dests {
-		dests[i] = d.Snapshot()
-	}
-	return RouteSnapshot{conf.Matcher, dests, "sendFirstMatch", route.key}
+	return makeSnapshot(&route.baseRoute, "sendFirstMatch")
+}
+
+func (route *RouteConsistentHashing) Snapshot() RouteSnapshot {
+	return makeSnapshot(&route.baseRoute, "consistentHashing")
+}
+
+// baseConfigExtender is a function that takes a baseRouteConfig and returns
+// a configuration object that implements RouteConfig. This function may be
+// the identity function, i.e., it may simply return its argument.
+// This mechanism supports maintaining different configuration objects for
+// different route types and creating a route-appropriate configuration object
+// on a configuration change.
+// The baseRoute object implements private methods (addDestination,
+// delDestination, etc.) that create a new baseRouteConfig object that reflects
+// the configuration change. This baseRouteConfig object is applicable to all
+// route types. However, some route types, like ConsistentHashingRoute, have
+// additional configuration and therefore have a distinct configuration object
+// that embeds baseRouteConfig (in the case of ConsistentHashingRoute, the object
+// is consistentHashingRouteConfig). This route-type-specific configuration also
+// needs to be updated on a base configuration change (e.g., on a change
+// affecting destinations). Accordingly, the public entry points that effect the
+// configuration change (Add, DelDestination, etc.) are implemented for baseRoute
+// and also for any route type, like ConsistentHashingRoute, that creates a
+// configuration object. These public entry points call the private method,
+// passing in a callback function of type baseConfigExtender, which takes a
+// baseRouteConfig and either returns it unchanged or creates an outer
+// configuration object with the baseRouteConfig embedded in it.
+// The private method then stores the RouteConfig object returned by the callback.
+type baseConfigExtender func(baseRouteConfig) RouteConfig
```

Comment:
interesting approach.  i was thinking maybe this a bit too much indirection/complication, i think i would just have the internal methods return the baseRouteConfig and have all the public methods (for example `RouteConsistentHashing.DelDestination`) take that return value and do any "config post processing", and then store it. this would admittedly have a bit more repetitive code but it might be easier to understand.

either way, i'm not opposed to your approach.  if you like it, it's fine with me.